Parameterize HttpApi base errorSchema

Avaq opened the pull request:
<!--
Before submitting a Pull Request, please ensure you've done the following:

- 📖 Read our Contributing Guide: https://github.com/effect-ts/.github/blob/main/CONTRIBUTING.md
- 📖 Read our Code of Conduct: https://github.com/effect-ts/.github/blob/main/CODE_OF_CONDUCT.md
- 👷‍♀️ Create small PRs. In most cases this will be possible.
- 📝 Use descriptive commit messages.
- ✅ Provide tests for your changes if applicable.
- 📗 If your change requires documentation, please update the relevant documentation.
- 📝 Create a changeset for your changes. This helps in tracking and communicating the changes effectively.
- ⏳ Please be patient! We will do our best to review your pull request as soon as possible.

NOTE: Pull Requests from forked repositories will need to be reviewed by a team member before any CI builds will run.
-->

## Type

<!--
What type of change is this? Please check all applicable.
-->

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This allows users to customise how HttpApiDecodeErrors are represented on the wire, enabling the use of HttpApi for wire protocols that don't follow the Tagged Error approach to encoding errors.

## Related

<!--
For pull requests that relate or close an issue, please include them below. We like to follow [Github's guidance on linking issues to pull requests](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

For example having the text: "closes #1234" would connect the current pull request to issue 1234 and automatically
close the issue once we merge the pull request.
-->

- Closes #5465

Avaq committed:
Parameterize HttpApi base errorSchema

This allows users to customize how HttpApiDecodeErrors are represented
on the wire, enabling the use of HttpApi for wire protocols that don't
follow the Tagged Error approach to encoding errors.

gcanti commented:
I'm not sure this will work as intended, in `handlerToRoute` within `HttpApiBuilder.ts` there's a `HttpApiDecodeError.refailParseError`, which makes me think the error logic is fixed

Avaq commented on packages/platform/src/HttpApi.ts:
```diff
@@ -255,14 +255,20 @@ const makeProto = <Id extends string, Groups extends HttpApiGroup.HttpApiGroup.A
-export const make = <const Id extends string>(identifier: Id): HttpApi<Id, never, HttpApiDecodeError> =>
+export const make = <const Id extends string>(
+  identifier: Id,
+  errorSchema: Schema.Schema<HttpApiDecodeError, any> = HttpApiDecodeError
```
If I try to use a discarded generic, TypeScript issues:

```
Type 'typeof HttpApiDecodeError' is not assignable to type 'Schema<HttpApiDecodeError, E, never>'.
Types of property 'Encoded' are incompatible.
Type '{ readonly _tag: "HttpApiDecodeError"; readonly message: string; readonly issues: readonly { readonly _tag: "Type" | "Pointer" | "Unexpected" | "Missing" | "Composite" | "Refinement" | "Transformation" | "Forbidden"; readonly message: string; readonly path: readonly (string | ... 1 more ... | { ...; })[]; }[]; }' is not assignable to type 'E'.
'E' could be instantiated with an arbitrary type which could be unrelated to '{ readonly _tag: "HttpApiDecodeError"; readonly message: string; readonly issues: readonly { readonly _tag: "Type" | "Pointer" | "Unexpected" | "Missing" | "Composite" | "Refinement" | "Transformation" | "Forbidden"; readonly message: string; readonly path: readonly (string | ... 1 more ... | { ...; })[]; }[]; }'. (ts 2322)
```

It's because allowing a user to supply `E` may invalidate the default parameter we're trying to assign.

Avaq replied:
[Updated](https://github.com/Effect-TS/effect/compare/69e05994d72f55a29420cf4008c64d9f8dbd9395..fc20fc2b86bb1234d6aac437fb853b242024237a) to use an options object

Avaq committed:
Fix the incorrect TypeScript type of httpApp (partially)

The HttpRouter.toHttpApp returns an App.Default<E | RouteNotFound>,
but that type gets suppressed by the input type of Router.router, which
has `unknown` for the `E`, so the whole thing become `unknown`.

Additionally, the `apiMiddleware` has the type:

    <E, R>(self: App.Default<E, R>): App.Default<any, any>

So once again, the RouteNotFound is lost when the router is piped
through the middleware.

Both of these places where type information is lost could have led to
the following oversight:

The error is eventually captured by catchAllCause, and an attempt is
made to encode it. We don't know if the encoding succeeds, but by
default it will not succeed for RouteNotFound, causing it to be
re-failed to the end user. It is then likely caught by the
`withErrorHandling` part of `HttpApp.toHandled` which uses
`ServerError.causeResponse` to convert it to a Response.

However, in userland it's not clear why this is happening, because the
return type from `httpApp` states that the created HttpApp cannot fail.

This commit adds RouteNotFound to the error branch of the returned
HttpApp, making it clearer for users how RouteNotFound is propegated and
handled, and allowing users to add their own handlers.

------------------------------------------------------------------------

Note:

Technically, all errors that could occur according to the injected HttpApi
should be included in the output type of the httpApp. We never know for
sure that the encoder will succeed, and so we could always re-fail them.

The thing is, we can't know these in the `httpApp` constructor, because
the `HttpApi` is injected via the `HttpApi.Api` tag, which loses the `E`
type on the `HttpApi` it contains. So I left this out of scope for this
fix, because it would require a lot of refactoring.

Avaq commented:
@tim-smart I added a new commit here because I found another issue. If you want, I can extract it to a separate PR. I guess these pulls don't get a lot of attention because you're all focused on smol. I'm looking forward to the day I can upgrade, and focus my contributions there.

fubhy commented:
Closing as stale. There has not been activity for more than 6 months, the PR is not merge-ready, and it has drifted in scope with an unrelated second commit. If this is still needed, please open a fresh focused PR with a changeset and current tests.
